src/odemis/driver/smaract.py

```diff
@@ -3173,3 +3174,1239 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        returns (float) the value
+        """
+        ret_val = c_double()
+        encoded_key = self.core.SA_SI_EPK(property_key, idx0, idx1)
+        self.core.SA_SI_GetProperty_f64(self._id, c_uint32(encoded_key), byref(ret_val), c_size_t(0))
+        return ret_val.value
+
+    def GetProperty_i32(self, property_key, idx0=0, idx1=0):
+        """
+        property_key (int32): property key symbol
+        idx0 (int): index value, meaning depends on property key
+        idx1 (int): index value, meaning depends on property key
+        returns (int) the value
+        """
+        ret_val = c_int32()
+        encoded_key = self.core.SA_SI_EPK(property_key, idx0, idx1)
+        self.core.SA_SI_GetProperty_i32(self._id, c_uint32(encoded_key), byref(ret_val), c_size_t(0))
+        return ret_val.value
+
+    def GetProperty_i64(self, property_key, idx0=0, idx1=0):
+        """
+        property_key (int64): property key symbol
+        idx0 (int): index value, meaning depends on property key
+        idx1 (int): index value, meaning depends on property key
+        returns (int) the value
+        """
+        ret_val = c_int64()
+        encoded_key = self.core.SA_SI_EPK(property_key, idx0, idx1)
+        self.core.SA_SI_GetProperty_i64(self._id, c_uint32(encoded_key), byref(ret_val), c_size_t(0))
+        return ret_val.value
+
+    def GetProperty_s(self, property_key, idx0=0, idx1=0):
+        """
+        property_key (int32): property key symbol
+        idx0 (int): index value, meaning depends on property key
+        idx1 (int): index value, meaning depends on property key
+        returns (str): the value
+        """
+        ret_val = create_string_buffer(SA_SI_STRING_MAX_LENGTH)
+        encoded_key = self.core.SA_SI_EPK(property_key, idx0, idx1)
+        self.core.SA_SI_GetProperty_s(self._id, c_uint32(encoded_key), ret_val, byref(c_size_t(len(ret_val))))
+        return ret_val.value.decode("latin1")
+
+    def GetValue_f64(self, channel, data_source_idx):
+        """
+        property_key (int32): property key symbol
+        data_source_idx:
+        returns (float) the value
+        """
+        ret_val = c_double()
+        self.core.SA_SI_GetValue_f64(self._id, channel, data_source_idx, byref(ret_val))
+        return ret_val.value
+
+    def WaitForEvent(self, timeout=600):
+        """
+        Blocks until device reports an event.
+        timeout (int > 0): timeout in seconds
+        returns (SA_SI_Event): event
+        raises
+            TimeoutError: timeout exceeded
+            CancelledError: function was cancelled by SA_SI_Cancel
+            SA_SI_Error: something went wrong
+        """
+        if timeout == float("inf"):
+            t = SA_SI_TIMEOUT_INFINITE
+        else:
+            t = c_uint(int(timeout * 1000))  # SA_SI_WaitForEvent accepts timeout in ms
+
+        ev = SA_SI_Event()
+        ret = self.core.SA_SI_WaitForEvent(self._id, byref(ev), t)
```

Comment:
That's not going to work, because `sp_errcheck()` already checks the return code and will raise an exception. So instead, your code should look like:
```python
try:
   self.core.SA_SI_WaitForEvent(self._id, byref(ev), t)
except SA_SIError as ex:
    if ex.errno == SA_SI_ERROR_TIMEOUT:
       raise TimeoutError(...)
    elif ex.errno == SA_SI_ERROR_CANCELLED:
       raise CancelledError(...)
    raise
```